### PR TITLE
[SYCL] Add SYCL registered kernel metadata into property sets

### PR DESCRIPTION
**Pull request by jzc**
For the sycl kernel compiler extension, see https://github.com/intel/llvm/pull/11985, https://github.com/intel/llvm/pull/16485

### CODE REVIEW

File: llvm/lib/SYCLLowerIR/ComputeModuleRuntimeInfo.cpp [hunk 1]
```diff
@@ -471,6 +471,20 @@ PropSetRegTy computeModuleProperties(const Module &M,
     }
   }
 
+  if (const NamedMDNode *MD = M.getNamedMetadata("sycl_registered_kernels")) {
+    assert(MD->getNumOperands() == 1 &&
+           "Invalid sycl_registered_kernels metadata");
+    const MDNode *RegistredKernels = MD->getOperand(0);
```
Comment (maarquitos14) [inline llvm/lib/SYCLLowerIR/ComputeModuleRuntimeInfo.cpp:477]:
```suggestion
    const MDNode *RegisteredKernels = MD->getOperand(0);
```

File: llvm/lib/SYCLLowerIR/ComputeModuleRuntimeInfo.cpp [hunk 2]
```diff
@@ -471,6 +471,20 @@ PropSetRegTy computeModuleProperties(const Module &M,
     }
   }
 
+  if (const NamedMDNode *MD = M.getNamedMetadata("sycl_registered_kernels")) {
+    assert(MD->getNumOperands() == 1 &&
```
Comment (maarquitos14) [inline llvm/lib/SYCLLowerIR/ComputeModuleRuntimeInfo.cpp:475]: Just asking: can it not be empty? I know it doesn't make a lot of sense, but just wondering if we really need an assert here or having it empty wouldn't be a big deal.

Comment (maarquitos14) [reply llvm/lib/SYCLLowerIR/ComputeModuleRuntimeInfo.cpp:475]: Btw, if it's valid, then we should add a new test with this scenario.

File: llvm/test/tools/sycl-post-link/sycl-registered-kernels.ll [hunk 3]
```diff
@@ -0,0 +1,41 @@
+; This test checks that the sycl-post-link ouputs registered kernel data 
+; from !sycl_registered_kernels metadata into the SYCL/registerd_kernels section.
```
Comment (maarquitos14) [inline llvm/test/tools/sycl-post-link/sycl-registered-kernels.ll:2]:
```suggestion
; from !sycl_registered_kernels metadata into the [SYCL/registered kernels] section.
```

File: llvm/test/tools/sycl-post-link/sycl-registered-kernels.ll [hunk 4]
```diff
@@ -0,0 +1,41 @@
+; This test checks that the sycl-post-link ouputs registered kernel data
```
Comment (maarquitos14) [inline llvm/test/tools/sycl-post-link/sycl-registered-kernels.ll:1]: Can we also test negative cases? If we finally decide to not accept empty metadata, then test that scenario. Same if we find a pair missing one element.

Comment (maarquitos14) [reply llvm/test/tools/sycl-post-link/sycl-registered-kernels.ll:1]: I also see in https://github.com/intel/llvm/pull/11985 the following:
```
It is not an error to have duplicate names in a
`registered_kernel_names` property, but the duplicates have no effect.
```

Can we have a test for this too?

**Comment by jopperm** [comment]
LGTM from the SYCL-RTC side, thanks!

**Comment by maarquitos14** [comment]
We also need at least an E2E test. It might be added in a follow up PR, but we need it.

**Comment by jopperm** [comment]
I observed that after device code split, the resulting modules always contain all `registered_kernel_names`. Could we emit only the entries whose RHS exist in the current module? (Maybe this needs to be handled in the module splitter. There seems to be precedent [here](https://github.com/intel/llvm/blob/20d4bec7e6413c1f6350981e420931ea1b8457ea/llvm/lib/SYCLLowerIR/ModuleSplitter.cpp#L303).)

**Comment by jzc** [comment]
@jopperm good point, I've updated the PR with the functionality to do this, take another look

File: llvm/test/tools/sycl-post-link/sycl-registered-kernels.ll (hunk 4 above)
Comment (maarquitos14) [reply llvm/test/tools/sycl-post-link/sycl-registered-kernels.ll:1]: I see tests for empty, and incorrect data format, but I think we still need testing for duplicates.

**Review by maarquitos14 — APPROVED** [review]
LGTM. Just a reminder that we need a follow up PR with E2E testing.

File: llvm/test/tools/sycl-post-link/sycl-registered-kernels-empty.ll [hunk 5]
```diff
@@ -0,0 +1,3 @@
+; RUN: sycl-post-link -properties %s -o %t.table
+; RUN: FileCheck %s -input-file=%t_0.prop --implicit-check-not="[SYCL/registered kernels]"
+!sycl_registered_kernels = !{}
```
Comment (jopperm) [inline llvm/test/tools/sycl-post-link/sycl-registered-kernels-empty.ll:3]: Nit: Missing newline.

**Review by jopperm — APPROVED** [review]
Thanks for adding the filtering!

File: llvm/test/tools/sycl-post-link/sycl-registered-kernels-empty.ll (hunk 5 above)
Comment (steffenlarsen) [reply llvm/test/tools/sycl-post-link/sycl-registered-kernels-empty.ll:3]:
```suggestion
!sycl_registered_kernels = !{}

```

**Comment by jopperm** [comment]
@jzc @steffenlarsen E2E failures seem unrelated and anecdotally (#17032) go away after merging with tip -- can you update the PR and land it, please?
